### DISCUSSION TITLE
Next.js server components and live preview

### DISCUSSION THREAD

**Discussion by bm-vs** [vercel/next.js#50360] [App Router]
I'm migrating a website to use the Next.js app/ directory and ran into some issues.
I'm using Sanity as the CMS for the website.

To render a component that requires fetching data I use 3 components: a server (entry point) component, a client (draft) component and a render component.

The server component is the entry point. It uses the [draftMode()](https://nextjs.org/docs/app/building-your-application/configuring/draft-mode#step-2-update-page) utility function.
If draft mode is disabled, it fetches the data that is needed and passes it to the render component. For example, fetching all of the content of a page.
Otherwise, it hands control over to the client component.

The client component uses [usePreview](https://github.com/sanity-io/next-sanity#next-sanitypreview-live-real-time-preview) to fetch the data and passes it to the render component.

The render component is just a dumb component that receives data and renders it. [It inherits the render context of its parent](https://nextjs.org/docs/getting-started/react-essentials#the-use-client-directive), so it's a Client Component when included by the draft component, and a Server Component when rendered by the entry point component.
Since it can be used in both contexts, both using server specific things like draftMode() and client specific things like useState() fails.

This pattern works well at the page level, but starts to have problems with nesting.
For example, including an ArticleList section that fetches the articles for itself inside of a Page.

How can the ArticleList component figure out if it's in draft mode or not? It's possible that it's being used either as a Client Component or as a Server Component.

The trivial solution is to use props, since that works both for Server Components and Client Components. But that doesn't scale very well. If there are multiple levels of components that fetch data, then I have to propagate isEnabled from draftMode() all the way from the last component that is exclusively a Server Component - the Page.

**draftMode()** can't be used, since that only works in the context of the server.
**useContext** is also excluded, since that only works on the client.
Another solution is to wrap **window** in a try/catch. This is what I'm doing and that seems to work well but also feels like a very hacky thing to do.

**Is there any alternative approach to doing this? Is there anything that I'm missing?**

Here's some code if it makes more sense:

page.tsx
```typescript
export default async function Page({params: {slug}}: Props) {
	const {isEnabled: isDraft} = draftMode();
	const pageSlug = slug.join('/');
	const pageQuery = groq`
		*[_type == "page" && slug.current == $slug && !(_id in path('drafts.**'))][0] {
			...
		}
	`;
	if (isDraft) {
		return <PageDraft pageQuery={pageQuery} pageSlug={pageSlug} />;
	}

	const page = await sanityFetch(pageQuery, PageSchema, {slug: pageSlug});
	return <PageRender page={page} />;
}
``` 

page.draft.tsx
```typescript
export const PageDraft: FC<Props> = ({pageQuery, pageSlug}) => {
	return (
		<PreviewSuspense fallback="Loading...">
			<Preview pageQuery={pageQuery} pageSlug={pageSlug} />
		</PreviewSuspense>
	);
};

const Preview: FC<Props> = ({pageQuery, pageSlug}) => {
	const page = usePreview(null, pageQuery, {slug: pageSlug});
	return <PageRender page={page} />;
};
```

page.render.tsx
```typescript
export const Page = ({page}: Props) => {
	const {content, heading} = page;
	const sections = content.map((section) => {
		switch (section._type) {
			case 'articleListAll':
				/* @ts-expect-error Async Server Component */
				return <ArticleListAll section={section} />;
			default:
				return <div>unknown</div>;
		}
	});

	return (
		<Suspense>
			<Heading activeHeading={heading.at(0)} />
			{sections}
		</Suspense>
	);
};
```

article-list-all.tsx
```typescript
export const ArticleListAll = async ({section}: Props) => {
	const isDraft = isClientContext();
	const articlesQuery = groq`
	*[_type == "article" && defined(slug.current) && !(_id in path('drafts.**'))]{
		...
	}`;
	if (isDraft) {
		return <ArticleListAllDraft section={section} articlesQuery={articlesQuery} />;
	}

	const articles = await sanityFetch(articlesQuery, z.array(ArticleSchema));
	return <ArticleListAllRender section={section} articles={articles} />;
};
```

article-list-all.draft.tsx
```typescript
export const ArticleListAllDraft: FC<Props> = ({section, articlesQuery}) => {
	return (
		<PreviewSuspense fallback="Loading...">
			<Preview section={section} articlesQuery={articlesQuery} />
		</PreviewSuspense>
	);
};

const Preview: FC<Props> = ({section, articlesQuery}) => {
	const articles = usePreview(null, articlesQuery);
	return <ArticleListAllRender section={section} articles={articles} />;
};
```

article-list-all.render.tsx
```typescript
export const ArticleListAllRender = ({section, articles}: Props) => {
	const {theme, title} = section;
	return null;
};
```

https://gist.github.com/BSantos22/85260e5129ad0ce1805202761644e626

**Comment by oogwayjay** [comment]
Hey @BSantos22, did you ever find a good solution for this? I'm running into the same thing now wondering if you made any progress outside of this discussion.

**Reply by bm-vs** [reply]
Not a good solution, no. And if you're using Sanity, and want to hop on the new Presentation tool, I think it's actually impossible at the moment.

**Reply by bm-vs** [reply]
Here's a more recent discussion about this: sanity-io/visual-editing#827

**Reply by markoleavy** [reply]
I'm actually using sanity with old `Iframe` preview, and I get real time editing. I'm not using usePreview in next.

**Comment by markoleavy** [comment]
I don't know if it's what your looking for, but what I'm doing right now is moving all off the queries on server components and conditionally call different queries/endpoints (and different cache-settings), based on the `isEnabled` variable.
I then wrap server pages in a client preview wrapper to get the page updated.
```javascript
// page.tsx
const Page = async ({
  params,
}: {
  params: { tour: string };
}) => {
   const { isEnabled } = draftMode();
   let props;
   if (isEnalbed){
      props = fetch('draftEndPointAndVariables',{ cache: 'no-store' })
   } else {
      props = fetch('prodEndpoitAndVariables')
   }
   const page = <PageComponent props={props}/>
   if (isEnabled){
      return <PreviewWrapper page={page}/>
   } else {
      return page
   }
}
```

```javascript
'use client';
import React, { useEffect } from 'react';

import { useRouter } from 'next/navigation';

const PreviewWrapper: React.FC<{ page: React.ReactNode }> = ({ page }) => {
  const router = useRouter();
  useEffect(() => {
    setInterval(() => {
      router.refresh();
    }, 10000);
  });
  return page;
};

export default PreviewWrapper;
```
